libs/SalesforceHybridSDK/SalesforceHybridSDK/Classes/Plugins/SFNetworkPlugin/SFNetworkPlugin.m

```diff
@@ -124,7 +124,8 @@ - (void) pgSendRequest:(CDVInvokedUrlCommand *) command
     [restApiInstance sendRequest:request
                                       failureBlock:^(id response, NSError *e, NSURLResponse *rawResponse) {
                                           __strong typeof(self) strongSelf = weakSelf;
-                                          CDVPluginResult *pluginResult = [CDVPluginResult resultWithStatus:CDVCommandStatus_ERROR messageAsString:e.localizedDescription];
+                                          NSString *error = [SFJsonUtils JSONRepresentation:response];
```

Comment:
So, if we pass back `error` which is basically `response`, what about `e`? Could we add `e` to the JSON we send back to the error callback, i.e. some structure like this:
```
{
    response: <response>,
    error: <e>
}
```
What do you think, @wmathurin?

Reply:
That's a good idea. That would make it almost backwards compatible.
If we return
```
{ 
  "response": <responseJson>, 
  "error": <e.localizedDescription>
}
```
And they have code to make sense of the existing error, they could keep it as long as they go again the error field.

Reply:
👍

Reply:
@bbirman Could we do the same on Android (https://github.com/forcedotcom/SalesforceMobileSDK-Android/blob/dev/libs/SalesforceHybrid/src/com/salesforce/androidsdk/phonegap/plugin/SalesforceNetworkPlugin.java#L168)? That would address https://github.com/forcedotcom/SalesforceMobileSDK-CordovaPlugin/issues/505.

Reply:
Started looking into this... I'm confused about where we'd have a response and an exception, the response is [returned here ](https://github.com/forcedotcom/SalesforceMobileSDK-Android/blob/0d5cb019fed36548febc7656e1509b4bce8dbcab/libs/SalesforceHybrid/src/com/salesforce/androidsdk/phonegap/plugin/SalesforceNetworkPlugin.java#L128) if the error code isn't a successful one. It looks like `onError` is [called here](https://github.com/forcedotcom/SalesforceMobileSDK-Android/blob/0d5cb019fed36548febc7656e1509b4bce8dbcab/libs/SalesforceHybrid/src/com/salesforce/androidsdk/phonegap/plugin/SalesforceNetworkPlugin.java#L162) if there's an issue with parsing, is that where we want to have the error also include the response?

Reply:
Should we build a little error dictionary like we do on iOS?

Instead of doing
```java
callbackContext.error(response.asString()); // when we have a non-2xx response
callbackContext.error(exception.getMessage()); // when we don't have a response
```
We could do return
```
{ 
  "response": <response.asString() if available>, 
  "error": <e.getMessage() if there is one>,
  "urlResponse": <response.getRawResponse().toString()>
}
```
Also maybe we don't need response and url response, the latter probably contains the former. So the final shape could be:
```
{
  response: { headers: , statusCode:, body: }, // if available
  error: e // if available
}
```

Reply:
We should probably do the same for `ReactNative` too, for consistency.